Mathlib/Tactic/Congrm.lean

```diff
@@ -0,0 +1,137 @@
+`congrm e` takes an expression `e` containing placeholders `_` and scans `e, lhs, rhs` in parallel.
+
+It matches both `lhs` and `rhs` to the pattern `e`, and produces one goal for each placeholder,
+stating that the corresponding subexpressions in `lhs` and `rhs` are equal.
+
+Examples:
+```lean
+example {a b c d : ℕ} :
+  Nat.pred a.succ * (d + (c + a.pred)) = Nat.pred b.succ * (b + (c + d.pred)) := by
+  congrm Nat.pred (Nat.succ _) * (_ + _)
+/-  Goals left:
+⊢ a = b
+⊢ d = b
+⊢ c + a.pred = c + d.pred
+-/
+  sorry
+  sorry
+  sorry
+
+example {a b : ℕ} (h : a = b) : (λ y : ℕ => ∀ z, a + a = z) = (λ x => ∀ z, b + a = z) := by
+  congrm λ x => ∀ w, _ + a = w
+  -- produces one goal for the underscore: ⊢ a = b
+  exact h
+```
+-/
+syntax (name := congrM) "congrm " term : tactic
+
+section util
+
+/-- Given a function, extract the explicit arguments. -/
+def _root_.Lean.Expr.getExplicitArgs (f : Expr) : MetaM (Array Expr) := do
+  let args := f.getAppArgs
+  let pinfo := (← Lean.Meta.getFunInfo f.getAppFn').paramInfo
+  return (pinfo.zip args).filterMap (λ arg => if arg.1.isExplicit then some arg.2 else none)
+
+/-- try to close the goal using refl, return the `mvarId` if it fails-/
+def tryRefl (goal : MVarId) : MetaM (Option MVarId) := do
+  let res ← observing? do
+    goal.refl
+  if res == none then return some goal else return none
+
+/-- to be moved somewhere else-/
+def _root_.Lean.MVarId.applyWithFreshMVarLevels (goal : MVarId) (lem : Name) :
+    MetaM (List MVarId) := do
+  goal.apply (← mkConstWithFreshMVarLevels lem)
+
+end util
+
+private partial def telescopingFn (lem : Name) (goal : MVarId) (x : Expr) : MetaM MVarId := do
+  let userName ← x.fvarId!.getUserName
+  let newGoals ← goal.applyWithFreshMVarLevels lem
+  if newGoals.length == 1 then
+    let newGoal := newGoals[0]!
+    return (← newGoal.intro userName).2
+  throwTacticEx `congrm goal m!"failed to apply {lem}"
+
+open private applyCongrThm? from Lean.Meta.Tactic.Congr in
+
+/-- Main loop for `congrm` -/
+partial def congrmLoop (pat : Expr) (goal : MVarId) : MetaM (List MVarId) := do
+  -- Helper function (stolen from somewhere) that creates the correct FVars in `λ` and `∀`
+  -- and does the recursion
+  let binders (mvarId : MVarId) (n : Name) (xs : Array Expr) (k : Expr) : MetaM (List MVarId) := do
+    congrmLoop k (← xs.foldlM (telescopingFn n) mvarId)
+  if pat.isMVar then
+    return [goal]
+  else if pat.isForall then
+    trace[congrm] s!"Forall pattern: {← ppExpr pat}"
+    forallTelescope pat (binders goal ``pi_congr)
+  else if pat.isLambda then
+    trace[congrm] s!"Lambda pattern: {← ppExpr pat}"
+    lambdaTelescope pat (binders goal ``funext)
+  else if pat.isApp then
+    let patternArgs := (← pat.getExplicitArgs).toList
+    trace[congrm] s!"Apply pattern, fun: {← ppExpr pat.getAppFn}, args: {← patternArgs.mapM ppExpr}"
+    let some congrThm ← mkCongrSimp? pat.getAppFn' (subsingletonInstImplicitRhs := false) |
+      return []
```

Comment:
Does this need an error message? If we haven't consumed the pattern yet the user probably should know about it!